syntax = "proto3";

package university;

option java_package = "com.university.grpc";
option go_package = "universitypb";
//option python_package = "university_pb";

// ---- ENUM Tan覺m覺 ----
enum LoanStatus {
  ONGOING = 0;
  RETURNED = 1;
  LATE = 2;
}

// ---- Message Tan覺mlar覺 ----

message Book {
  string id = 1;
  string title = 2;
  string author = 3;
  string isbn = 4;
  string publisher = 5;
  int32 pageCount = 6;
  int32 stock = 7;
}

message Student {
  string id = 1;
  string name = 2;
  string studentNumber = 3;
  string email = 4;
  bool isActive = 5;
}

message Loan {
  string id = 1;
  string studentId = 2;
  string bookId = 3;
  string loanDate = 4;
  string returnDate = 5;
  LoanStatus status = 6;
}

// Ortak mesajlar
message IdRequest {
  string id = 1;
}
message Empty {}

message BookList { repeated Book books = 1; }
message StudentList { repeated Student students = 1; }
message LoanList { repeated Loan loans = 1; }

// ---- Servisler ----

service BookService {
  rpc ListBooks(Empty) returns (BookList);
  rpc GetBook(IdRequest) returns (Book);
  rpc AddBook(Book) returns (Book);
  rpc UpdateBook(Book) returns (Book);
  rpc DeleteBook(IdRequest) returns (Empty);
}

service StudentService {
  rpc ListStudents(Empty) returns (StudentList);
  rpc GetStudent(IdRequest) returns (Student);
  rpc AddStudent(Student) returns (Student);
  rpc UpdateStudent(Student) returns (Student);
  rpc DeleteStudent(IdRequest) returns (Empty);
}

service LoanService {
  rpc ListLoans(Empty) returns (LoanList);
  rpc GetLoan(IdRequest) returns (Loan);
  rpc BorrowBook(Loan) returns (Loan);
  rpc ReturnBook(IdRequest) returns (Loan);
}
